src/big/big_decimal.cr

```diff
@@ -0,0 +1,286 @@
+
+    i = 0
+    while remainder != ZERO && i < max_div_iterations
+      inner_quotient, inner_remainder = remainder.divmod(denominator)
+      quotient = quotient * TEN + inner_quotient
+      remainder = inner_remainder * TEN
+      i += 1
+    end
+
+    BigDecimal.new(normalize_quotient(other, quotient), scale + i)
+  end
+
+  def <=>(other : BigDecimal) : Int32
+    if @scale > other.scale
+      @value <=> other.scale_to(self).value
+    elsif @scale < other.scale
+      scale_to(other).value <=> other.value
+    else
+      @value <=> other.value
+    end
+  end
+
+  def <=>(other : Int)
+    self <=> BigDecimal.new(other)
+  end
+
+  def ==(other : BigDecimal) : Bool
+    if @scale > other.scale
+      scaled = other.value * power_ten_to(@scale - other.scale)
+      @value == scaled
+    elsif @scale < other.scale
+      scaled = @value * power_ten_to(other.scale - @scale)
+      scaled == other.value
+    else
+      @value == other.value
+    end
+  end
+
+  # Scales a `BigDecimal` to another `BigDecimal`, so they can be
+  # computed easier.
+  def scale_to(new_scale : BigDecimal) : BigDecimal
+    new_scale = new_scale.scale
+
+    if @value == 0
+      BigDecimal.new(0.to_big_i, new_scale)
+    elsif @scale > new_scale
+      scale_diff = @scale - new_scale.to_big_i
+      BigDecimal.new(@value / power_ten_to(scale_diff), new_scale)
+    elsif @scale < new_scale
+      scale_diff = new_scale - @scale.to_big_i
+      BigDecimal.new(@value * power_ten_to(scale_diff), new_scale)
+    else
+      self
+    end
+  end
+
+  def to_s(io : IO)
+    s = @value.to_s
+
+    if @scale == 0
+      io << s
+      return
+    end
+
+    if @scale == s.size
+      io << "0." << s
+    elsif @scale > s.size
+      io << "0."
+      (@scale - s.size).times do
+        io << '0'
+      end
+      io << s
+    else
+      offset = s.size - @scale
+      io << s[0...offset] << '.' << s[offset..-1]
+    end
+  end
+
+  def to_big_d
+    self
+  end
+
+  def clone
+    self
+  end
+
+  private def check_division_by_zero(bd : BigDecimal)
+    raise DivisionByZero.new if bd.value == 0
+  end
+
+  private def power_ten_to(x : Int) : Int
+    TEN ** x
+  end
+
+  # Returns the quotient as absolutely negative if self and other have different signs,
+  # otherwise returns the quotient.
+  private def normalize_quotient(other : BigDecimal, quotient : BigInt) : BigInt
+    if (@value < 0 && other.value > 0) || (other.value < 0 && @value > 0)
+      -quotient.abs
+    else
+      quotient
+    end
+  end
+end
+
+struct Int
+  include Comparable(BigDecimal)
+
+  # Convert `Int` to `BigDecimal`
+  def to_big_d
+    BigDecimal.new(self)
+  end
+
+  def <=>(other : BigDecimal)
+    self <=> other.value
+  end
+end
+
+class String
+  include Comparable(BigDecimal)
+
+  # Convert `String` to `BigDecimal`
+  def to_big_d
+    BigDecimal.new(self)
+  end
+end
+
+struct Float
+  # Casting from `Float` is not recommended due to precision loss risks.
```

Comment:
I'd suggest adding separate `NOTE` paragraph about precision loss.

Reply:
Modified the comment here and above in `initialize` a bit.